File: src/platform/macos/window.m
#import "utils.m"
#import <Cocoa/Cocoa.h>
#import <QuartzCore/QuartzCore.h>
#import <WebKit/WebKit.h>

struct WPoint {
    uint16 x;
    uint16 y;
};

struct WSize {
    uint16 width;
    uint16 height;
};

struct WRect {
    struct WSize size;
    struct WPoint origin;
};

static NSWindowStyleMask styleMaskFromTag(uint8_t tag) {
    switch (tag) {
    case 0:
        return NSWindowStyleMaskBorderless;
    case 1:
        return NSWindowStyleMaskTitled;
    case 2:
        return NSWindowStyleMaskClosable;
    case 3:
        return NSWindowStyleMaskMiniaturizable;
    case 4:
        return NSWindowStyleMaskResizable;
    case 5:
        return NSWindowStyleMaskUnifiedTitleAndToolbar;
    case 6:
        return NSWindowStyleMaskFullScreen;
    case 7:
        return NSWindowStyleMaskFullSizeContentView;
    default:
        return 0;
    }
}

void setTitle(void *window_ptr, const char *title) {
    NSString *title_string = [[NSString alloc] initWithUTF8String:title];
    NSWindow *window = windowFromPtr(window_ptr);
    [window setTitle:title_string];
}

void setRect(void *window_ptr, struct WRect *rect, bool animated, bool display) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setFrame:(NSMakeRect(rect->origin.x, rect->origin.y, rect->size.width, rect->size.height))
             display:(display)animate:(animated)];
}

void getRect(void *window_ptr, struct WRect *out) {
    NSWindow *window = windowFromPtr(window_ptr);
    out->size.width = window.frame.size.width;
    out->size.height = window.frame.size.height;
    out->origin.x = window.frame.origin.x;
    out->origin.y = window.frame.origin.y;
}

void insertStyleMask(void *window_ptr, uint8_t tag) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setStyleMask:([window styleMask] | styleMaskFromTag(tag))];
}

void removeStyleMask(void *window_ptr, uint8_t tag) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setStyleMask:([window styleMask] & ~styleMaskFromTag(tag))];
}

void setTitleVisibility(void *window_ptr, bool is_visible) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setTitleVisibility:(is_visible ? NSWindowTitleVisible : NSWindowTitleHidden)];
}

void setTitleBarAppearsTransparent(void *window_ptr, bool is_appears_transparent) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setTitlebarAppearsTransparent:is_appears_transparent];
}

void setMovableByWindowBackground(void *window_ptr, bool is_movable) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setMovableByWindowBackground:is_movable];
}

void setWindowBackgroundColor(void *window_ptr, double r, double g, double b, double a) {
    NSWindow *window = windowFromPtr(window_ptr);
    NSColor *color = [NSColor colorWithSRGBRed:r green:g blue:b alpha:a];
    [window setBackgroundColor:color];
    [window setOpaque:(a >= 1.0)];
}

// --- Window Constraints ---

void setMinSize(void *window_ptr, struct WSize *size) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setMinSize:NSMakeSize(size->width, size->height)];
}

void setMaxSize(void *window_ptr, struct WSize *size) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setMaxSize:NSMakeSize(size->width, size->height)];
}

// --- Window Visibility ---

void showWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window makeKeyAndOrderFront:nil];
}

void hideWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window orderOut:nil];
}

void focusWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window makeKeyAndOrderFront:nil];
    if (@available(macOS 14.0, *)) {
        [NSApp activate];
    } else {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
        [NSApp activateIgnoringOtherApps:YES];
#pragma clang diagnostic pop
    }
}

bool isWindowVisible(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    return [window isVisible];
}

// --- Window State ---

void minimizeWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window miniaturize:nil];
}

void maximizeWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window zoom:nil];
}

void closeWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window performClose:nil];
}

void toggleFullScreen(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    if (!([window styleMask] & NSWindowStyleMaskFullScreen)) {
        [window toggleFullScreen:nil];
    }
}

void restoreWindow(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    if ([window isMiniaturized]) [window deminiaturize:nil];
    if ([window styleMask] & NSWindowStyleMaskFullScreen) [window toggleFullScreen:nil];
    if ([window isZoomed]) [window zoom:nil];
}

// --- Window Level/Ordering ---

void setAlwaysOnTop(void *window_ptr, bool on_top) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window setLevel:(on_top ? NSFloatingWindowLevel : NSNormalWindowLevel)];
}

void orderWindowFront(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window orderFront:nil];
}

void orderWindowBack(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);
    [window orderBack:nil];
}

// --- Window Dragging ---

// Persistent monitor captures every mouseDown so we have the NSEvent
// available when the async bind() callback fires from the webview.
static id mouseDownMonitor = nil;
static NSEvent *lastMouseDownEvent = nil;

void startDragging(void *window_ptr) {
    NSWindow *window = windowFromPtr(window_ptr);

    // Lazily install a persistent mouseDown monitor
    if (!mouseDownMonitor) {
        mouseDownMonitor = [NSEvent addLocalMonitorForEventsMatchingMask:NSEventMaskLeftMouseDown
                                                                handler:^NSEvent *(NSEvent *event) {
            lastMouseDownEvent = event;
            return event;
        }];
    }

    // Use the captured mouseDown event, fall back to currentEvent
    NSEvent *event = lastMouseDownEvent ?: [NSApp currentEvent];
    lastMouseDownEvent = nil; // Clear after use to prevent stale drags
    if (event && event.type == NSEventTypeLeftMouseDown) {
        [window performWindowDragWithEvent:event];
    }
}

void removeDragMonitor(void) {
    if (mouseDownMonitor) {
        [NSEvent removeMonitor:mouseDownMonitor];
        mouseDownMonitor = nil;
    }
    lastMouseDownEvent = nil;
}

File: src/platform/macos/Window.zig
const Self = @This();
const objc = @import("extern.zig");
const Rect = @import("../../models/Rect.zig");
const Size = @import("../../models/Size.zig");

const DEFAULT_TRAFFIC_LIGHTS_SPACING: f64 = 22;

pub const StyleMask = enum(u8) {
    borderless,
    titled,
    closable,
    miniaturizable,
    resizable,
    unified_title_and_toolbar,
    full_screen,
    full_size_content_view,
};

handle: *anyopaque,

pub fn init(handle: *anyopaque) Self {
    return .{ .handle = handle };
}

pub fn setTitle(self: *const Self, title: [*:0]const u8) void {
    objc.setTitle(self.handle, title);
}

pub fn getRect(self: *const Self) Rect {
    var rect: Rect = undefined;
    objc.getRect(self.handle, &rect);
    return rect;
}

pub fn setRect(self: *const Self, rect: Rect, animated: bool, display: bool) void {
    objc.setRect(self.handle, &rect, animated, display);
}

pub fn setStyleMask(self: *const Self, style_mask: []const StyleMask) void {
    for (style_mask) |mask| {
        objc.insertStyleMask(self.handle, @intFromEnum(mask));
    }
}

pub fn setTitleVisibility(self: *const Self, is_visible: bool) void {
    objc.setTitleVisibility(self.handle, is_visible);
}

pub fn setTitleBarAppearsTransparent(self: *const Self, is_appears_transparent: bool) void {
    objc.setTitleBarAppearsTransparent(self.handle, is_appears_transparent);
}

pub fn setTrafficLightsPosition(self: *const Self, x: f64, y: f64, spacing: ?f64) void {
    objc.setTrafficLightsPosition(
        self.handle,
        x,
        y,
        spacing orelse DEFAULT_TRAFFIC_LIGHTS_SPACING,
    );
}

pub fn setMovableByWindowBackground(self: *const Self, is_movable: bool) void {
    objc.setMovableByWindowBackground(self.handle, is_movable);
}

pub fn setBackgroundColor(self: *const Self, r: f64, g: f64, b: f64, a: f64) void {
    objc.setWindowBackgroundColor(self.handle, r, g, b, a);
}

pub fn setWebViewTransparent(self: *const Self) void {
    objc.setWebViewTransparent(self.handle);
}

// Window constraints
pub fn setMinSize(self: *const Self, size: Size) void {
    objc.setMinSize(self.handle, &size);
}

pub fn setMaxSize(self: *const Self, size: Size) void {
    objc.setMaxSize(self.handle, &size);
}

// Window visibility
pub fn show(self: *const Self) void {
    objc.showWindow(self.handle);
}

pub fn hide(self: *const Self) void {
    objc.hideWindow(self.handle);
}

pub fn focus(self: *const Self) void {
    objc.focusWindow(self.handle);
}

pub fn isVisible(self: *const Self) bool {
    return objc.isWindowVisible(self.handle);
}

// Window state
pub fn minimize(self: *const Self) void {
    objc.minimizeWindow(self.handle);
}

pub fn maximize(self: *const Self) void {
    objc.maximizeWindow(self.handle);
}

pub fn close(self: *const Self) void {
    objc.closeWindow(self.handle);
}

pub fn fullscreen(self: *const Self) void {
    objc.toggleFullScreen(self.handle);
}

pub fn restore(self: *const Self) void {
    objc.restoreWindow(self.handle);
}

// Window level/ordering
pub fn setAlwaysOnTop(self: *const Self, on_top: bool) void {
    objc.setAlwaysOnTop(self.handle, on_top);
}

pub fn orderFront(self: *const Self) void {
    objc.orderWindowFront(self.handle);
}

pub fn orderBack(self: *const Self) void {
    objc.orderWindowBack(self.handle);
}

// Window dragging
pub fn startDragging(self: *const Self) void {
    objc.startDragging(self.handle);
}
